Use Json TypeInfoResolverChain

eerhardt opened the pull request:
- Removes all usages of JsonSerializerOptions.AddContext, which will be obsoleted by https://github.com/dotnet/runtime/issues/83280
- Update ProblemDetailsJsonContext to be added in a `Configure<JsonOptions>()` call back and to always be added to the beginning of the resolver chain at that time
    - This gives us the simplest, most understandable pattern for all libraries to follow.
- Small clean up in the API template's usings

eerhardt committed:
Use Json TypeInfoResolverChain

- Removes all usages of JsonSerializerOptions.AddContext, which will be obsoleted by https://github.com/dotnet/runtime/issues/83280
- Update ProblemDetailsJsonContext to be added just before the DefaultJsonTypeInfoResolver if it is the last resolver in the chain. Otherwise, add it to the end of a non-empty resolver chain.
- Small clean up in the API template's usings

BrennanConroy commented on src/Http/Http.Extensions/src/ProblemDetailsJsonOptionsSetup.cs:
```diff
@@ -11,23 +11,27 @@ internal sealed class ProblemDetailsJsonOptionsSetup : IPostConfigureOptions<Jso
     {
-        switch (options.SerializerOptions.TypeInfoResolver)
+        var typeInfoResolverChain = options.SerializerOptions.TypeInfoResolverChain;
+        if (typeInfoResolverChain.Count == 0)
```
Hmm, I know this isn't new in this PR, but should we instead add ours and the reflection based one in this case?

Reason is because this configure call is specifically to add our ProblemDetailsJsonContext, but now because there aren't any resolvers you also get the reflection based resolver for Problem Details, feels wrong. Especially if the normal use case for getting no resolvers is when trimming is on, which is when you would want our provided context even more.

eerhardt replied:
After much off-GitHub discussion, we'd decided on the following approach:

1. `Microsoft.AspNetCore.Http.Json.JsonOptions` sets the `TypeInfoResolver` to `DefaultJsonTypeInfoResolver` at construction-time based on the STJ feature switch allowing Reflection or not. If Reflection isn't allowed, it leaves it null.  (This is how it works today - but based on the current ASP.NET feature switch).
2. `AddProblemDetails()` registers a `Configure<JsonOptions>()` call back, and always adds the `ProblemDetailsJsonContext` to the beginning of the chain at that time

This has the following behaviors:

1. In a "normal" CoreCLR app, you get the chain `{ ProblemDetailsContext, DefaultResolver }`, like you want
2. If someone clears/nulls etc any of this after the fact, the behavior they get is determined by what they changed. At least it is deterministic and understandable.
3. In a trimmed/aot app, we will only add `{ ProblemDetailsContext }`, and it is up to the app to add in their SG JsonContext, which can be done before or after the PD context, as they need
    - When someone doesn't, there will be no JsonTypeInfo for their types, and they will get the same behavior they do today
4. This sets a pattern for other libraries who require the same behavior (adding their own source generated JsonContext to the resolver chain).

davidfowl commented on src/Http/Http.Extensions/src/ProblemDetailsJsonOptionsSetup.cs:
```diff
@@ -11,23 +11,27 @@ internal sealed class ProblemDetailsJsonOptionsSetup : IPostConfigureOptions<Jso
+            return;
+        }
+
+        var lastResolver = typeInfoResolverChain[typeInfoResolverChain.Count - 1];
```
```suggestion
        var lastResolver = typeInfoResolverChain[^1];
```

eiriktsarpalis commented on src/Http/Http.Extensions/src/ProblemDetailsJsonOptionsSetup.cs:
```diff
@@ -11,23 +11,27 @@ internal sealed class ProblemDetailsJsonOptionsSetup : IPostConfigureOptions<Jso
+        }
+
+        var lastResolver = typeInfoResolverChain[typeInfoResolverChain.Count - 1];
+        if (lastResolver is DefaultJsonTypeInfoResolver)
```
This should account for any user customization as well:
```suggestion
        if (lastResolver is DefaultJsonTypeInfoResolver defaultResolver && 
            defaultResolver.GetType() == typeof(DefaultJsonTypeInfoResolver) && 
            defaultResolver.Modifiers.Count == 0)
```

eiriktsarpalis commented on src/Http/Http.Extensions/src/ProblemDetailsJsonOptionsSetup.cs:
```diff
@@ -11,23 +11,27 @@ internal sealed class ProblemDetailsJsonOptionsSetup : IPostConfigureOptions<Jso
+            // In this case, the current configuration has a reflection-based resolver at the end
+            // and we are inserting our internal ProblemDetails context to be evaluated
+            // just before the reflection-based resolver.
+            typeInfoResolverChain.Insert(typeInfoResolverChain.Count - 1, ProblemDetailsJsonContext.Default);
```
Shouldn't this method account for the possibility that `JsonSerializerOptions` has been made read-only by the user? I believe @brunolins16 was using the copy constructor elsewhere to guard against that case.

eerhardt replied:
Use of the copy constructor was added in #45906 to work around https://github.com/dotnet/runtime/issues/80527. See the conversation [here](https://github.com/dotnet/aspnetcore/pull/45906/commits/2540eb41a7329ba0c378e6327ad13fb31a135e55#r1068458691).

That was undone in #46716 once the fix for https://github.com/dotnet/runtime/issues/80527 came in. See https://github.com/dotnet/aspnetcore/issues/46143.

The user shouldn't make the options read-only until all the configuration has run. Them trying to do so, and then us modifying it anyways, isn't really a good solution IMO. Note that the only way they can make the options read-only is if they make it read-only in a `Configure<JsonOptions>()` callback - there's no way for them to access this object earlier. Making it read-only in the middle of configuration doesn't make sense because other configuration needs to run.

eerhardt commented on src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:
```diff
@@ -9,7 +9,7 @@
 builder.Services.ConfigureHttpJsonOptions(options =>
 {
-    options.SerializerOptions.AddContext<AppJsonSerializerContext>();
+    options.SerializerOptions.TypeInfoResolverChain.Insert(0, AppJsonSerializerContext.Default);
```
It isn't really about the "default" type resolver to use though. For example if you call `AddProblemDetails()`, you don't want to use this type resolver for the ProblemDetails types.

I assume you'd be OK with the readability of:

```C#
options.SerializerOptions.TypeInfoResolverChain.Add(AppJsonSerializerContext.Default);
```

But that's not the semantics of what you want. You really want to "prepend" by default. So if we were to add something to the framework, I think it would be a helper to add a context to the beginning of the chain.

> could we perhaps create an extension method on Microsoft.AspNetCore.Http.Json.JsonOptions

If we want something in the framework, I believe it would be better to put it in System.Text.Json itself.

eerhardt replied:
Other alternatives here are:

1. Use Combine with the new one first:
```C#
options.SerializerOptions.TypeInfoResolver =
    JsonTypeInfoResolver.Combine(AppJsonSerializerContext.Default, options.SerializerOptions.TypeInfoResolver);
```
2. Since this is only for NativeAOT, we could just use `.Add` and rely on the fact that when in NativeAOT there won't be a reflection based resolver. However, when you aren't using NativeAOT, you don't want to use `.Add` because that will put your context _after_ the reflection based resolver, which means yours won't be used. So I wouldn't go with this approach.

DamianEdwards replied:
My thought process was that setting the "default" type resolver in the chain means putting it first. We can leave this as is for now and we can consider if there's a higher level API that could help with the usability here later.

eerhardt replied:
Would the "Combine" API be easier on the eyes? They are equivalent.

DamianEdwards replied:
Possibly, although it has the problem of being quite long. Let's leave the `Insert(0` for now and revisit.

halter73 replied:
I was worried using `JsonTypeInfoResolver.Combine` would collapse the `TypeInfoResolverChain` giving it a length of one making it less inspectable, but then I looked at the `AddFlattened` logic and saw that it should indeed be equivalent. I agree it `Combine` looks nicer than `Insert(0`.

eerhardt committed:
Change ProblemDetailsJsonOptionsSetup to use Configure instead of PostConfigure.

When adding the ProblemDetailsJsonContext, we always prepend it to the beginning of the chain at the time the configure step is executed, no matter the state of the chain.

This allows for a simpler, more understandable policy for all libraries that want to add their JsonContext into the JsonSerializerOptions resolver chain. The order of the chain is now determined by the order that the configure steps were registered in DI (for example when AddProblemDetails() was called).

halter73 commented on src/Http/Http.Extensions/src/ProblemDetailsServiceCollectionExtensions.cs:
```diff
@@ -40,7 +40,8 @@ public static IServiceCollection AddProblemDetails(this IServiceCollection servi
         services.TryAddEnumerable(ServiceDescriptor.Singleton<IProblemDetailsWriter, DefaultProblemDetailsWriter>());
-        services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<JsonOptions>, ProblemDetailsJsonOptionsSetup>());
+        // Use IConfigureOptions (instead of post-configure) so the registration gets added/invoked relative to when AddProblemDetails() is called.
+        services.TryAddEnumerable(ServiceDescriptor.Singleton<IConfigureOptions<JsonOptions>, ProblemDetailsJsonOptionsSetup>());
```
Nit: Now that it's normal configure, it's more common just to combine it with the user-specified callback if any. something like.

```csharp
services.Configure<ProblemDetailsOptions>(options =>
{
    options.SerializerOptions.TypeInfoResolver =
        JsonTypeInfoResolver.Combine(AppJsonSerializerContext.Default, options.SerializerOptions.TypeInfoResolver);

    configure?.Invoke(options);
});
```

eerhardt replied:
That breaks this test that is explicitly testing that duplicates aren't added:

https://github.com/dotnet/aspnetcore/blob/138388372bb7c6e35fdfa679da740fabc275a688/src/Http/Http.Extensions/test/ProblemDetailsServiceCollectionExtensionsTest.cs#L33-L47

halter73 replied:
And I was conflating two different options types lol.

eerhardt commented:
FYI - I plan on merging this today to unblock the runtime work in https://github.com/dotnet/runtime/pull/84022. Let me know if you have further feedback.
